docs/fundamentals/gateway/custom-gateway.md

```diff
@@ -0,0 +1,328 @@
+(custom-gateway)=
+# Customize the Gateway
+A Jina Gateway is customizable and can be implemented in the same way Executors are implemented.
+With Custom Gateways, Jina gives power back to the users, by allowing to implement any server, protocol and interface on
+the gateway level. This means you have more freedom in:
+* Choosing your favourite server framework.
+* Choosing the protocol used to serve your app.
+* Defining your API Gateway interface. You can define your JSON schema or protos,...
+
+Customization is allowed different components:
+* Implementing the custom gateway using a `base` gateway class: {class}`~jina.Gateway` or {class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`.
+* Using the {class}`~jina.serve.streamer.GatewayStreamer` to send data to Executors in the Flow.
+* Implementing the needed health-checks for the runtime.
+* Optionally, define a `config.yml` for it.
+* Optionally, bootstrapping the gateway using a `Dockerfile` and re-using the docker image.
+
+## Implementing the custom gateway
+Similarly to how you would implement an Executor, you can implement a custom gateway by inheriting from a base gateway class.
+Jina Gateway Runtime will instantiate your implemented class, inject runtime arguments and user-defined arguments to it, 
+run it, send health-checks to it and orchestrate it.
+
+Two Base Gateway classes are provided to allow implementing a custom gateway:
```

Comment:
Lets add FastAPI only in a separatr section

Reply:
there's already a separate subsection for fast api